src/Nethermind/Nethermind.Merge.AuRa/Shutter/ShutterP2P.cs

```diff
@@ -0,0 +1,145 @@
+using System.Threading.Tasks;
+using System.Threading;
+using Microsoft.Extensions.DependencyInjection;
+using System.Collections.Generic;
+using Multiformats.Address;
+using Nethermind.Consensus.AuRa.Config;
+using Nethermind.Logging;
+using ILogger = Nethermind.Logging.ILogger;
+using System.Threading.Channels;
+
+namespace Nethermind.Merge.AuRa.Shutter;
+
+public class ShutterP2P
+{
+    private readonly Action<Dto.DecryptionKeys> _onDecryptionKeysReceived;
+    private readonly ILogger _logger;
+    private readonly Channel<byte[]> _msgQueue = Channel.CreateUnbounded<byte[]>();
+    private readonly IAuraConfig _auraConfig;
+    private PubsubRouter? _router;
+    private ServiceProvider? _serviceProvider;
+    private CancellationTokenSource? _cancellationTokenSource;
+
+    public ShutterP2P(Action<Dto.DecryptionKeys> onDecryptionKeysReceived, IAuraConfig auraConfig, ILogManager logManager)
+    {
+        _onDecryptionKeysReceived = onDecryptionKeysReceived;
+        _logger = logManager.GetClassLogger();
+        _auraConfig = auraConfig;
+    }
+
+    public void Start(in IEnumerable<string> p2pAddresses)
+    {
+        _serviceProvider = new ServiceCollection()
+            .AddLibp2p(builder => builder)
+            .AddSingleton(new IdentifyProtocolSettings
+            {
+                ProtocolVersion = _auraConfig.ShutterP2PProtocolVersion,
+                AgentVersion = _auraConfig.ShutterP2PAgentVersion
+            })
+            .AddSingleton(new Settings()
+            {
+                ReconnectionAttempts = int.MaxValue
+            })
+            // .AddLogging(builder =>
+            //     builder.SetMinimumLevel(Microsoft.Extensions.Logging.LogLevel.Trace)
+            //     .AddSimpleConsole(l =>
+            //     {
+            //         l.SingleLine = true;
+            //         l.TimestampFormat = "[HH:mm:ss.FFF]";
+            //     })
+            // )
+            .BuildServiceProvider();
+
+        IPeerFactory peerFactory = _serviceProvider.GetService<IPeerFactory>()!;
+        ILocalPeer peer = peerFactory.Create(new Identity(), "/ip4/0.0.0.0/tcp/" + _auraConfig.ShutterP2PPort);
+        if (_logger.IsInfo) _logger.Info($"Started Shutter P2P: {peer.Address}");
+        _router = _serviceProvider.GetService<PubsubRouter>()!;
+
+        ITopic topic = _router.Subscribe("decryptionKeys");
+
+        topic.OnMessage += (byte[] msg) =>
+        {
+            _msgQueue.Writer.TryWrite(msg);
+            if (_logger.IsDebug) _logger.Debug($"Received Shutter P2P message.");
+        };
+
+        MyProto proto = new();
+        _cancellationTokenSource = new();
+        _ = _router.RunAsync(peer, proto, token: _cancellationTokenSource.Token);
+        ConnectToPeers(proto, p2pAddresses);
+
+        long lastMessageProcessed = DateTimeOffset.Now.ToUnixTimeSeconds();
+        long delta = 0;
+
+        Task.Run(() =>
+        {
+            for (; ; )
+            {
+                Thread.Sleep(250);
```

Comment:
You do not need to Sleep when using a channel. Readers are notified immediately and only when there is something to consume.